Fixes for 1.20.3

mat-1 opened the pull request:
This PR has two fixes:
- chatcomponent (aka Component.Serializer) doesn't have any strings since 23w46a, so I changed it to be identified by the field and method signatures, and checking that it doesn't have any constant strings.
- Some block/item ids were changed to be defined in references.Blocks and references.Items classes, so I added identifiers for those classes and made it extract the strings from the disassembly.

I tested with 1.14.4, 23w40a, and 1.20.3-pre2.

mat-1 committed:
fix chatcomponent detection for 23w46a

mat-1 committed:
handle references class so melons/pumpkins work again

mat-1 committed:
don't error on older versions

mat-1 committed:
remove debug print lol

mat-1 committed:
fix for 1.20.3-pre2 lol

Pokechu22 commented on burger/toppings/identify.py:
```diff
@@ -105,49 +104,50 @@ def check_match(value, match_list):
 
-    for c in classloader.search_constant_pool(path=path, type_=String):
+    class_file = classloader[path]
+
+    string_constants = classloader.search_constant_pool(path=path, type_=String)
+    for c in string_constants:
```
`classloader.search_constant_pool` is a performance optimization to avoid having to parse the whole class file for every class in the game.

Using `time python3 munch.py -t identify -o /dev/null 23w45a.jar`, on my main machine the original code takes about 2.5s to execute before and 5s after. On the Raspberry Pi 2 I use to automatically run Burger for new snapshots, it takes about 65s to execute before and 115s to execute after.

I don't think this is an acceptable slowdown; we should only parse the whole class file if needed. This could maybe be done by removing the `type_=String` part and instead checking with `isinstance`; the Gson field check could be done in the same way.

mat-1 replied:
Ah, I made the mistake of assuming search_constant_pool returned a list (which is the reason my chatcomponent detection broke). I also didn't realize how slow classloader was, I just made a commit that should result in it having similar performance to how it was before.

Pokechu22 commented on burger/toppings/items.py:
```diff
@@ -259,6 +276,14 @@ def on_get_field(self, ins, const, obj):
+                    if const.name_and_type.name.value in references_class_fields_to_item_ids:
+                        return references_class_fields_to_item_ids[const.name_and_type.name.value]
+                    else:
+                        print("Unknown field", const.name_and_type.name.value, "in references class", references_class)
```
should check `if verbose`

Pokechu22 reviewed (changes requested):
This doesn't seem to work for me; `chatcomponent` doesn't seem to be identified in 23w46a/1.20.3-pre1/1.20.3-pre2; although it no longer causes identify to error out and prevent other stuff from running, it does break the packets topping... though the packets topping is already somewhat broken due to #49.

mat-1 committed:
better chatcomponent detection and check verbose for a couple prints

mat-1 committed:
optimize chatcomponent identification

mat-1 committed:
optimize some more

i forgot to commit this earlier, whoops

mat-1 commented:
I addressed your review a couple days ago btw, I'm replying to generate a notification in case you didn't see.

Pokechu22 commented:
I saw but haven't had time to look into it further yet.

Are you planning on also looking into #49 (either in this PR or in another one)? I have a vague idea of how it could be fixed (once we identify packetbuffer, scan through it to identify all of the functions that just call a single netty function and build a dictionary of those, and then use that when decompiling) but haven't had time to implement it.

mat-1 commented:
> Are you planning on also looking into #49 (either in this PR or in another one)?

In this PR, no, but I might attempt to fix it later if I find the motivation and energy to. For now I'm just doing the bare minimum to keep my client up-to-date with the latest Minecraft version.

Pokechu22 commented on burger/toppings/identify.py:
```diff
@@ -105,200 +104,250 @@ def check_match(value, match_list):
+                # This technique will not work in 1.8.
+                logger_type = "Lorg/apache/logging/log4j/Logger;"
+                while not cf.fields.find_one(type_=logger_type):
+                    if cf.super_.name == "java/lang/Object":
```
You ended up deleting `cf = classloader[path]` and then not re-adding it. This is fairly obvious if you view the diff with whitespace differences disabled (`git diff -w 1.19-dev` or an option on the gear on the github web interface). This only affects older versions though.

Pokechu22 commented on burger/toppings/identify.py:
```diff
@@ -105,200 +104,250 @@ def check_match(value, match_list):
+
+                methods = list(class_file.methods.find(f=is_serialize_method))
+                if len(methods) == 1:
+                    return "chatcomponent", methods[0].args[0].name
```
This check fails in 1.19 through 1.20.2. `if len(methods) > 0` works though, so I'll switch to that.

Pokechu22 reviewed (commented):
I found a few minor issues which I'll add an extra commit to handle. Performance for identify on my pi is now 70s, which is close enough to the original 65s; good job.

<details><summary>Versions I regenerated, for reference</summary>

```
1.7.10 1.8
1.8.9 1.9
1.8 1.10
1.8 1.8.9
1.8 1.9
1.9.4 1.10
1.10.2 1.11
1.11.2 1.12
1.12.2 1.13
1.13.2 1.14
1.14.4 1.15
1.15.2 1.16
1.16.5 1.17
1.17.1 1.18
1.18.2 1.19
1.19.4 1.20
1.20 1.20.1
1.20.1 1.20.2
1.20.2 23w40a
23w40a 23w41a
23w41a 23w42a
23w42a 23w43a
23w43a 23w43b
23w43b 23w44a
23w44a 23w45a
23w45a 23w46a
23w46a 1.20.3-pre1
1.20.3-pre1 1.20.3-pre2
1.20.3-pre2 1.20.3-pre3
1.20.3-pre3 1.20.3-pre4
```

</details>

From that, old versions all work correctly (after my fix), but identification of chatcomponent fails in 23w40a through 23w45a (it works in 23w46a and on). I'm going to merge this despite that flaw, since it's better to have some data for current versions, but it's something that should be fixed at some point.

Pokechu22 committed:
Fixes for older versions

len(methods) is more than 1 on 1.19 through 1.20.2.

cf = classloader[path] was accidentally deleted and then not re-added, which breaks some extremely old versions.

There was some trailing whitespace (that existed prior to these changes).
